app/src/main/java/com/rockthevote/grommet/ui/registration/RegistrationDataTransformer.kt

```diff
@@ -0,0 +1,353 @@
+
+import android.util.Base64
+import com.rockthevote.grommet.R
+import com.rockthevote.grommet.data.db.model.*
+import com.rockthevote.grommet.ui.registration.address.AddressData
+import com.rockthevote.grommet.ui.registration.name.PersonNameData
+import com.rockthevote.grommet.util.Dates
+import java.util.*
+
+/**
+ * Handles transformation of [registrationData]
+ */
+class RegistrationDataTransformer @Throws(InvalidRegistrationException::class) constructor(
+    private val registrationData: RegistrationData,
+    private val partnerInformation: PartnerInformation
+) {
+
+    init {
+        validateRegistrationData(registrationData)
+    }
+
+    // These have already be evaluated as non null in [validateRegistrationData]
+    private val newRegistrationData = registrationData.newRegistrantData!!
+    private val addressData = registrationData.addressData!!
+    private val additionalInfoData = registrationData.additionalInfoData!!
+    private val assistanceData = registrationData.assistanceData!!
+    private val reviewData = registrationData.reviewData!!
+
+    /**
+     * Transforms the [registrationData] and [partnerInformation] into a [RockyRequest]
+     *
+     * @throws [InvalidRegistrationException] if unable to validate [RegistrationData]
+     */
+    fun transform(
+        unknown: UnknownDataSource
+    ): RockyRequest {
+
+        val body = RockyRequestBody(
+            lang = reviewData.formLanguage,
+            phoneType = null, // TODO I can't find where/what is supposed to populate this
+            partnerId = partnerInformation.partnerId,
+            optInEmail = additionalInfoData.hasOptedIntoNewsUpdates,
+            optInSms = additionalInfoData.hasOptedIntoNewsCallAndText,
+            optInVolunteer = additionalInfoData.hasOptedForVolunteerText,
+            // TODO All of these data sources are unknown, I think they may be from partner information, but don't know for sure
+            partnerOptInSms = unknown.partnerOptInSms,
+            partnerOptInEmail = unknown.partnerOptInEmail,
+            partnerOptInVolunteer = unknown.partnerOptInVolunteer,
+            finishWithState = true, // TODO is this a default value?
+            createdViaApi = true, // TODO is this a default value?
+            sourceTrackingId = unknown.sourceTrackingId,
+            partnerTrackingId = unknown.partnerTrackingId,
+            geoLocation = unknown.geoLocation,
+            openTrackingId = unknown.openTrackingId,
+            voterRecordsRequest = buildVoterRecordsRequest()
+        )
+
+        return RockyRequest(body)
+    }
+
+    private fun buildVoterRecordsRequest(): VoterRecordsRequest {
+        return VoterRecordsRequest(
+            type = "registration", // TODO is this a default value?
+            // TODO is this from somewhere else, I can't find a source?
+            // TODO Does this _have_ to be a full datetime stamp? It is in the example, but I don't see a formatter anywhere in the project
```

Comment:
most likely has to be a full time stamp (the API is very finicky about data types)

it's formatted [here](https://github.com/TrustTheVote-Project/Grommet/blob/6b0230af2e1f500db3943544cc64a177bd44d0ed/app/src/main/java/com/rockthevote/grommet/data/api/model/ApiVoterRecordsRequest.java#L44)

it's generated as soon as a new registration row is created, [here](https://github.com/TrustTheVote-Project/Grommet/blob/8c10e8b88a44b5e28c6161cf71685b75d519a05b/app/src/main/java/com/rockthevote/grommet/data/db/model/RockyRequest.java#L239). It's safe to say that we can generate this within the transformer (basically the time the voter finishes registering, so safe difference).
 I'm sure there was a reason for for putting default values in the various places (builder, sql, other?), just hard to remember why?

Reply:
Ah.... I see, IDK why, but I was looking at this as just a date rather than date/time

Reply:
I see why that was the case now, the example outputs date in the format:
`"generated_date": "2016-06-16T19:44:45+00:00"`

Whereas the function to do this outputs it as:
`"generated_date": "2020-04-11T19:50Z"`

The current settings match the previous code, but not the example.